### ISSUE TITLE
Misleading Error Message: the trait `Foo` is not implemented for `&dyn Foo`

### ISSUE THREAD

**Issue by ObliqueMotion** [rust-lang/rust#73492]
Here is a minimal example [[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3f1ec48f5216c6f700615bcd401333c8)]:

```rust
pub trait Foo {
    fn bar(&self, _: &dyn Foo) -> bool {
        true
    }
}

impl Foo for u32 {}

fn main() {
    let x = 1;
    let v: Vec<&dyn Foo> = vec![&1, &2, &3];
    v.iter().all(|y| x.bar(y));
}
```

Which produces this error:

```
error[E0277]: the trait bound `&dyn Foo: Foo` is not satisfied
  --> src/main.rs:14:28
   |
14 |     v.iter().all(|y| x.bar(y));
   |                            ^ the trait `Foo` is not implemented for `&dyn Foo`
   |
   = note: required for the cast to the object type `dyn Foo`
```

I find this error to be misleading because:

1) The type of `y` within the closure should be `&&dyn Foo`, not `&dyn Foo`.
2) Even though the type listed in the error message seems wrong, the trait `Foo` _should_ be implemented for `&dyn Foo` if that were the actual type of `y`.

This can be fixed in a number of ways:
```rust
v.iter().all(|&y| x.bar(y));
v.iter().all(|&y| x.bar(&*y));
v.iter().all(|y| x.bar(*y));
v.iter().all(|y| x.bar(&**y));
v.into_iter().all(|y| x.bar(y));
```
All of which are consistent with `y` being of type `&&dyn Foo`, and none of which are suggested by the error message.

**Comment by estebank** [comment]
```
error[E0277]: the trait bound `&dyn Foo: Foo` is not satisfied
  --> src/main.rs:14:28
   |
14 |     v.iter().all(|y| x.bar(y));
   |                            ^ the trait `Foo` is not implemented for `&dyn Foo`
   |
   = help: the trait `Foo` is implemented for `u32`
   = note: required for the cast from `&&dyn Foo` to `&dyn Foo`
```

It should suggest `*y`.

**Comment by estebank** [comment]
Triage: very little change.

```
rror[E0277]: the trait bound `&dyn Foo: Foo` is not satisfied
  --> src/main.rs:14:28
   |
14 |     v.iter().all(|y| x.bar(y));
   |                            ^ the trait `Foo` is not implemented for `&dyn Foo`
   |
help: the trait `Foo` is implemented for `u32`
  --> src/main.rs:7:1
   |
 7 | impl Foo for u32 {}
   | ^^^^^^^^^^^^^^^^
   = note: required for the cast from `&&dyn Foo` to `&dyn Foo`
```
